### ISSUE TITLE
validation_data parameter to fit with preprocessing

### ISSUE THREAD

**Issue by gioxc88** [adriangb/scikeras#131]
It's  pretty common in keras to pass `validation_data` to `fit` to monitor  how the loss behaves out of sample for each epoch.

I noticed that `scikeras` offers two solutions:

1. pass `validation_split` in the initialization 
2. pass `fit__validation_data` in the initialization 

Since I need to preprocess `X` and `y` before fitting I cannot use validation_split, because this would cause data leakage so I should opt for solution 2 but this makes the model stateful since the `fit__validation_data` is attached to the `model` instance.

Do you have any suggestion for this problem?

Many thanks

**Comment by adriangb** [comment]
This is indeed an interesting question.

Could you do:
```python3
est.set_params(fit__validation_data=...)
est.fit(...)
est.set_params(fit__validation_data=None)
```
?

But perhaps we should put `validation_data` back into `fit`; I believe it's the only parameter that suffers from this complication / is actual data.

Could you elaborate on how you are preprocessing your data and how that would lead to data leakage? Are you using SciKeras' data pre/post processing interface?

**Comment by gioxc88** [comment]
> This is indeed an interesting question.
> 
> Could you do:
> 
> ```python
> est.set_params(fit__validation_data=...)
> est.fit(...)
> est.set_params(fit__validation_data=None)
> ```
> 
> ?
> 
> But perhaps we should put `validation_data` back into `fit`; I believe it's the only parameter that suffers from this complication / is actual data.
> 
> Could you elaborate on how you are preprocessing your data and how that would lead to data leakage? Are you using SciKeras' data pre/post processing interface?

Thank you for the answer.
I am using `scikeras` as part of a bigger project and I always use `sklearn` like transformers to preprocess data inside a `pipeline`,  so basically `scikeras.KerasRegressor` instance would be last step of the `pipeline`.

Say for example (simple case) I want to use `StandardScaler` in the preprocessing step.
Passing `validation_split` means that I have a combined `[X_train, X_val]` dataset which contains both training and validation data. I apply preprocessing to the combined dataset and then pass it to `scikeras.KerasRegressor` which splits it according to `validation_split`. In this case both `X_train` and `X_val` are standardized using all data which causes leakage. 

Of course there are solutions to this, for example preprocess  `X_train` and `X_val` separeately, then combine them and figure out `validation_split = len(X_val) // (len(X_train) + len(X_val))`.

However I believe this workflow in a pipeline it's not easy (if not impossible) to do because all the elements in the `pipeline` shares the same `X` and `y` whereas the `validation_data` is a parameter specific to `keras`  so in a `pipeline` the proposed solution will not work.

I could easily subclass `KerasRegressor` and wrap the fit function like this:

```
class KerasRegressorMod(KerasRegressor)
    def fit(self, X, y, sample_weight=None, validation_data=None):
        self.fit__validation_data = validation_data
        super().fit(X, y, sample_weight)
        del self.fit__validation_data
        return self
```
I was wondering though if change the signature of fit directly in `scikeras` to allow `validation_data` to be passed directly.

Many thanks
Gio

**Comment by adriangb** [comment]
Thank you for the detailed explanation. I do agree that that looks like the best solution right now, but I'd like to explore the problem a bit more since I think that would constitute a divergence from the sklearn API.

Looking into how this problem is solved in other parts of the sklearn ecosystem:
* I could not find any sklearn estimator that accepts actual validation data, only splits.
* [skorch](https://skorch.readthedocs.io/en/latest/user/FAQ.html#i-already-split-my-data-into-training-and-validation-sets-how-can-i-use-them) seems to implement this the same way we do here (there are some implementation differences, but same idea).

What are the problems with the `fit__validation_data` approach? Obviously this attaches a reference to the data to the model, and I see many ways in which that could lead to problems, but I'd like to know what problems that creates for you exactly.

**Comment by gioxc88** [comment]
> Thank you for the detailed explanation. I do agree that that looks like the best solution right now, but I'd like to explore the problem a bit more since I think that would constitute a divergence from the sklearn API.
> 
> Looking into how this problem is solved in other parts of the sklearn ecosystem:
> 
> * I could not find any sklearn estimator that accepts actual validation data, only splits.
> * [skorch](https://skorch.readthedocs.io/en/latest/user/FAQ.html#i-already-split-my-data-into-training-and-validation-sets-how-can-i-use-them) seems to implement this the same way we do here (there are some implementation differences, but same idea).
> 
> What are the problems with the `fit__validation_data` approach? Obviously this attaches a reference to the data to the model, and I see many ways in which that could lead to problems, but I'd like to know what problems that creates for you exactly.

Thank you for taking the time to look into this.
The problem is both conceptual and practical:
1. conceptual simply because, as I said `sklearn` instances are stateless and attaching `validation_data` to the instance breaks this rule.
2. practical because there is a reason why `sklearn` instances are stateless and this is because if you call `fix` on a different `X, y` pair, you still get consistent results.
However, if you `fix` on a different `X, y` pair on `scikeras` models it is possible that `validation_data` is inconsistent with the new `X, y`. But since you attached it to the instance when you call fit the `model_.history` parameter will be wrong because the loss is calculated on a `validation_data` inconsistent with the new `X, y` and in my opinion this is a bad implementation.

My question is, why do you think that adding `validation_data` to fit would break the `sklearn` API? there are various estimator which accepts `**fit_params` and also in the official paper where the API is described it is never mentioned that `fit` should not accept additional parameters other that `X, y`

Many thanks
Gio

**Comment by adriangb** [comment]
I _feel_ like it might be breaking the API because I couldn't find _any_ sklearn estimator that accepts _data_ via `**fit_params`. But like you say, just because it hasn't been done doesn't mean that it's actually breaking the API, it just means that it should be carefully considered. Perhaps this is a good question to ask in the sklearn gitter?

**Comment by adriangb** [comment]
Wondering if you have any ideas @stsievert

**Comment by gioxc88** [comment]
> I _feel_ like it might be breaking the API because I couldn't find _any_ sklearn estimator that accepts _data_ via `**fit_params`. But like you say, just because it hasn't been done doesn't mean that it's actually breaking the API, it just means that it should be carefully considered. Perhaps this is a good question to ask in the sklearn gitter?

`Pipeline` for example does

**Comment by adriangb** [comment]
How does `Pipeline` accept data via fit params (aside from it just forwarding arbitrary data via parameter routing)?

**Comment by stsievert** [comment]
I don't like the fact that Keras's `fit` accepts validation data. That seems like a waste; I think that should be done by the user once per epoch with the `score` function or any of the relevant loss functions.

@gioxc88  why do you need to pass the estimator validation data? Why can't you score yourself once per epoch with your special preprocessing?

I'd remove `fit__validation_data` from the SciKeras API. Data definitely should not be attached to the estimator. If anything, validation data should be passed alongside the training data in `**fit_params` because it keeps the estimator functional. Yes, there's not a Scikit-learn estimator that follows this exact pattern, but I don't think that's a huge deal.

**Comment by gioxc88** [comment]
> I don't like the fact that Keras's `fit` accepts validation data. That seems like a waste; I think that should be done by the user once per epoch with the `score` function or any of the relevant loss functions.
> 
> @gioxc88 why do you need to pass the estimator validation data? Why can't you score yourself once per epoch with your special preprocessing?
> 
> I'd remove `fit__validation_data` from the SciKeras API. Data definitely should not be attached to the estimator. If anything, validation data should be passed alongside the training data in `**fit_params` because it keeps the estimator functional. Yes, there's not a Scikit-learn estimator that follows this exact pattern, but I don't think that's a huge deal.

Thank you for the answer.
Your solution would be ideal (call the loss function outside once per epoch) but to my knowledge it's not possible because you don't have access to the model weights of each epoch.

If I am wrong I would really appreciate if you could point me in the right direction.

Many thanks
Gio

**Comment by stsievert** [comment]
> but to my knowledge it's not possible because you don't have access to the model weights of each epoch.

I think you do – SciKeras exposes a `model_` attribute, which is a normal Keras model. So wouldn't this work?

``` python
from sklearn.metrics import mean_squared_error

est = KerasRegressor(...)
for epoch in range(100):
    est.partial_fit(X_train, y_train)  # one epoch

    y_pred = est.model_.predict(X_test)
    loss_test = mean_squared_error(y_test, y_pred)
    print(f"On epoch={epoch}, received test loss={loss_test}")
```

**Comment by adriangb** [comment]
Another alternative: [custom callback](https://www.google.com/search?q=keras+custom+callback&oq=keras+custom+callback&aqs=chrome..69i57j0l3j69i60l4.2842j0j7&sourceid=chrome&ie=UTF-8).

**Comment by gioxc88** [comment]
> > but to my knowledge it's not possible because you don't have access to the model weights of each epoch.
> 
> I think you do – SciKeras exposes a `model_` attribute, which is a normal Keras model. So wouldn't this work?
> 
> ```python
> from sklearn.metrics import mean_squared_error
> 
> est = KerasRegressor(...)
> for epoch in range(100):
>     est.partial_fit(X_train, y_train)  # one epoch
> 
>     y_pred = est.model_.predict(X_test)
>     loss_test = mean_squared_error(y_test, y_pred)
>     print(f"On epoch={epoch}, received test loss={loss_test}")
> ```

Of course I could do that but tbh I don't see any advantages of doing it like this over passing `validation_data` to fit. In fact I believe it's better to pass `**fit_params` than to  calculate the loss in a for loop, because you still need to train using partial fit every iteration. It would have been different if you could access the actual model at each epoch and just calculate the loss on the test set without needing to train.

**Comment by gioxc88** [comment]
> Another alternative: [custom callback](https://www.tensorflow.org/guide/keras/custom_callback).
> 
> Edit: I meant to link to the docs, not my google search for the docs, I did not mean to be rude.

How do you think this can be used as an alternative?

**Comment by adriangb** [comment]
I haven't tried it (I will at some point, maybe tomorrow) but you should be able to do something like:

```python3
x_val, y_val = ...
loss = ...

class CustomCallback(keras.callbacks.Callback):

    def on_epoch_end(self, epoch, logs=None):
        print("Epoch {}: {}".format(epoch, loss(self.model.predict(x_val), y_val))

est = KerasRegressor(fit__callbacks=[CustomCallback()])
````

I'm not sure that's any better, but it should be _an_ option.

**Comment by adriangb** [comment]
If we were to incorporate this as a fit kwarg, I wonder if we should support it via SciKeras' data pre/post processing interface. My gut instinct is that we should _not_, since that might add needless complexity.

**Comment by stsievert** [comment]
> we should not

Yes, I concur. There are good alternatives with `validation_frac` and the custom callbacks in https://github.com/adriangb/scikeras/issues/131#issuecomment-730009238.

**Comment by adriangb** [comment]
@gioxc88 does the custom callback work for your use case?

This still leaves the question of what to do with this parameter. It seems like it would be better off in `fit`. That is what [LightGBM](https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.LGBMRegressor.html#lightgbm.LGBMRegressor.fit) does, and it has been at least proposed in sklearn (https://github.com/scikit-learn/scikit-learn/issues/18748). The other option of course is to remove it completely (users could still use it, it would just be undocumented).

**Comment by stsievert** [comment]
> This still leaves the question of what to do with this parameter.

Which parameter? `validation_frac`?

**Comment by adriangb** [comment]
I was talking about `validation_data`, I thought it was one of the hardcoded params in `__init__` but it's not. So I guess the options are:
1. Add it to`__init__` (we agreed this is not a good option)
2. Add it to `fit` (we agree this might be a viable option)
3. Do nothing (this is the default option).

**Comment by stsievert** [comment]
> I was talking about validation_data.

Got it. Thanks for laying it out clearly.

I think (3) also requires documentation. I'd like to see a clear example notebook, and maybe even a callback implementation in `scikeras.utils`. I think that would resolve most or all advanced use cases, and the basic parameter `validation_frac` would resolve most basic use cases.

**Comment by adriangb** [comment]
> I'd like to see a clear example notebook

Not a problem, it'll just be a question of what notebook / topic this fits into best.

> And maybe even a callback implementation in `scikeras.utils`

A callback specific to the above use case? Or a general callback utility?

**Comment by gioxc88** [comment]
> @gioxc88 does the custom callback work for your use case?
> 
> This still leaves the question of what to do with this parameter. It seems like it would be better off in `fit`. That is what [LightGBM](https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.LGBMRegressor.html#lightgbm.LGBMRegressor.fit) does, and it has been at least proposed in sklearn ([scikit-learn/scikit-learn#18748](https://github.com/scikit-learn/scikit-learn/issues/18748)). The other option of course is to remove it completely (users could still use it, it would just be undocumented).

I am already using callbacks for other purposes and I am aware I could have used callbacks for this as well.

However it requires a lot more effort and the main drawback is that, for saving the output of the validation losses I would need to create a new  instance attribute on `self.model` which is definitely not good practice.

As I said as long as I can use `fit__validation_data`, even if undocumented, I am ok and the I issue can be closed because I would subclass `KerasRegressor` as I want to modify the signature of the function to accept `**fit_params`.

The reason why I opened the issue is because I think that `scikeras` would benefit from this change.

Keeping track of the loss on the hold out set vs training set is one of the most important aspects of any serious ML workflow which involves Neural Networks, as well as data leakage which very often disregarded. For this reason passing `validation_data` to fit should only be natural and very easy to do. 

Unfortunately I believe that both proposed solution don't add any advantages to just changing the signature of `fit` to accept `**fit_params`. I can only see disadvantages in using either callbacks or for loops + partial fit.

That being said, whatever you decide, thank you for considering my thoughts on this. I believe this is an excellent package!

Many thanks
Gio

**Comment by gioxc88** [comment]
@adriangb I saw in the last commit you added kwargs to fit and predict.
Thank you!!

**Comment by adriangb** [comment]
Yes, it should help for now! But it is meant to be a temporary backward compatibility measure, so I would leave this issue open until we figure out a permanent solution to this specific problem.

**Comment by gioxc88** [comment]
Can I ask why do you think this should only be considered a permanent solution?

The kwargs in or fit_params in wrappe is definitely a must for wrapper libraries.

**Comment by adriangb** [comment]
The main reason is that most of these parameters exist as `__init__` params in sklearn and we try to follow the sklearn API here. It also makes a lot of things (like `get_params`) easier internally. They are also properties of the model, not really of the data. `fit__validation_data` is obviously an exception since there is nothing like it in sklearn and it is also a property of the data.

**Comment by stsievert** [comment]
@gioxc88 would this be an acceptable solution?

Let's say you decided to use the Keras API:

``` python
est = KerasClassifier(...)
est.fit(X, y, batch_size=128)
```

This would issue a warning/raise an error:

> Passing keyword arguments to `fit` has been deprecated in SciKeras 0.2.1, and support will be removed be 1.0.0. To resolve this issue, set these keyword arguments at initialization:
>
> * Use `KerasClassifier(..., batch_size=128)`  to use `batch_size=128` for all functions
> * Use `KerasClassifier(..., fit__batch_size=128)` to only use `batch_size=128` for fit.

This would encourage this behavior:

``` python
est = KerasClassifier(..., fit__batch_size=128, predict__batch_size=1000)
est.fit(X, y)
y_pred = est.predict(X)
```

(I think `predict__batch_size=1000` makes a good default; presumably, that would only affect `predict`).

**Comment by adriangb** [comment]
Hi @gioxc88 , sorry to bother you again.

We are discussing implementing a feature that might help with your use case. This is not a replacement for `**kwargs`, merely another option for when those are not possible (grid search, cross validation, etc.). Would you be able to take a look at this [example implementation](https://www.adriangb.com/scikeras/refs/pull/167/merge/notebooks/DataTransformers.html#6.-Custom-validation-dataset) and/or the `DatasetTransformer` section in these [docs](https://www.adriangb.com/scikeras/refs/pull/167/merge/advanced.html#data-transformers)? Thank you!
